demos/ota/ota_demo_core_mqtt/ota_demo_core_mqtt.c

```diff
@@ -0,0 +1,1681 @@
+    {
+        /* Send MQTT CONNECT packet to broker. */
+        mqttStatus = MQTT_Connect( pMqttContext, &connectInfo, NULL, CONNACK_RECV_TIMEOUT_MS, &sessionPresent );
+
+        pthread_mutex_unlock( &mqttMutex );
+    }
+    else
+    {
+        LogError( ( "Failed to acquire mutex for executing MQTT_Connect"
+                    ",errno=%s",
+                    strerror( errno ) ) );
+    }
+
+    if( mqttStatus != MQTTSuccess )
+    {
+        returnStatus = EXIT_FAILURE;
+        LogError( ( "Connection with MQTT broker failed with status %s.",
+                    MQTT_Status_strerror( mqttStatus ) ) );
+    }
+    else
+    {
+        LogInfo( ( "MQTT connection successfully established with broker.\n\n" ) );
+    }
+
+    return returnStatus;
+}
+
+/*-----------------------------------------------------------*/
+
+static int establishConnection( void )
+{
+    int returnStatus = EXIT_FAILURE;
+
+    /* Set the pParams member of the network context with desired transport. */
+    networkContext.pParams = &opensslParams;
+
+    /* Attempt to connect to the MQTT broker. If connection fails, retry after
+     * a timeout. Timeout value will be exponentially increased till the maximum
+     * attempts are reached or maximum timeout value is reached. The function
+     * returns EXIT_FAILURE if the TCP connection cannot be established to
+     * broker after configured number of attempts. */
+    returnStatus = connectToServerWithBackoffRetries( &networkContext );
+
+    if( returnStatus != EXIT_SUCCESS )
+    {
+        /* Log error to indicate connection failure. */
+        LogError( ( "Failed to connect to MQTT broker %.*s.",
+                    AWS_IOT_ENDPOINT_LENGTH,
+                    AWS_IOT_ENDPOINT ) );
+    }
+    else
+    {
+        /* Establish MQTT session on top of TCP+TLS connection. */
+        LogInfo( ( "Creating an MQTT connection to %.*s.",
+                   AWS_IOT_ENDPOINT_LENGTH,
+                   AWS_IOT_ENDPOINT ) );
+
+        /* Sends an MQTT Connect packet using the established TLS session,
+         * then waits for connection acknowledgment (CONNACK) packet. */
+        returnStatus = establishMqttSession( &mqttContext );
+
+        if( returnStatus != EXIT_SUCCESS )
+        {
+            LogError( ( "Failed creating an MQTT connection to %.*s.",
+                        AWS_IOT_ENDPOINT_LENGTH,
+                        AWS_IOT_ENDPOINT ) );
+        }
+        else
+        {
+            LogDebug( ( "Success creating MQTT connection to %.*s.",
+                        AWS_IOT_ENDPOINT_LENGTH,
+                        AWS_IOT_ENDPOINT ) );
+
+            mqttSessionEstablished = true;
+        }
+    }
+
+    return returnStatus;
+}
+
+/*-----------------------------------------------------------*/
+
+static void disconnect( void )
+{
+    /* Disconnect from broker. */
+    LogInfo( ( "Disconnecting the MQTT connection with %s.", AWS_IOT_ENDPOINT ) );
+
+    if( mqttSessionEstablished == true )
```

Comment:
I have some concerns about the thread safety of this, as access to this variable isn't protected by a mutex. Also, the network disconnect may end up being called multiple times, as we don't check or keep track of how many times it's called